src/Invariants/Equality.jl

```diff
@@ -0,0 +1,21 @@
+using Graphs
```

Comment:
I like what you've done! However, once I started thinking about it, I realized what you really wanted to do was to _override_ `==`. This can be done because every operation in Julia is a function. So we write new methods for `==`. For example:

```julia 
import Base: ==
using Graphs.SimpleGraphs: SimpleEdge

# Define a helper function to normalize an edge. Notice that
# I prefix the function name with an underscore to indicate
# that it is a helper function and is not meant to be used
# outside of this file.
function _normalize_edge(edge)
    edge isa Tuple && return (min(edge...), max(edge...))
    return (min(edge.src, edge.dst), max(edge.src, edge.dst))
end

function ==(
    a::Union{Vector{Tuple{Int, Int}}, Vector{SimpleEdge{Int64}}},
    b::Union{Vector{Tuple{Int, Int}}, Vector{SimpleEdge{Int64}}},
)
    a_set = Set(_normalize_edge(edge) for edge in a)
    b_set = Set(_normalize_edge(edge) for edge in b)

    return a_set == b_set
end
```

Notice how we write a new method for `==` which acts on the objects we are concerned with.  Moreover, we will no longer need to export `equals` or in this case `==` from `GraphProperties.jl` since overloading base functions like == doesn't require exporting them, because they are already part of the Base module which is universally accessible. When you overload == for specific types, Julia will use your custom definition automatically when those types are compared using ==.

So, if you override == for particular types in your GraphProperties.jl script, you don't need to export ==. As long as the types for which you're overloading == are accessible (i.e., either defined in Base, imported from another module, or exported from your module), your overloaded method will be used.

Sorry for drawing out this PR, but I think doing this early on will make you a much better programmer in the long run. 

Also, I asked GPT4 to help write a good doc string (which I had to edit a little bit). The Julia example for the new method you can literally copy past into a REPL and check that it works. However, its not going to render well in GitHub and might look funny. Hopefully you can see what it should be 

```julia
"""
    ==(
        a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool

Determine if two sets of edges are equivalent in an undirected graph context.

This method treats each edge as undirected, meaning the order of vertices in
each edge tuple or `Graphs.SimpleGraphs.SimpleEdge` is ignored. Two sets of edges
are considered equal if they contain the same pairs of vertices, regardless of
the order in which the vertices appear in each edge.

# Arguments
- `a`: A vector of edges, either as tuples or `Graphs.SimpleGraphs.SimpleEdge{Int64}`.
- `b`: Another vector of edges to compare against `a`.

# Returns
- `true` if both sets of edges represent the same undirected connections, `false` otherwise.

# Examples
```julia
julia> using Graphs.SimpleGraphs: SimpleEdge

julia> using GraphProperties.Invariants

julia> edges_a = [(1, 2), (3, 4)]
2-element Vector{Tuple{Int64, Int64}}:
 (1, 2)
 (3, 4)

julia> edges_b = [SimpleEdge(2, 1), SimpleEdge(4, 3)]
2-element Vector{SimpleEdge{Int64}}:
 Edge 2 => 1
 Edge 4 => 3

julia> edges_a == edges_b
true
```
"""
```

Reply:
Would this override `==` for any 

```
a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}}
```
?

I'm curious if this is not what we want since we may want to compare other vectors of edges where order matters. For example, if we want to compare two paths, we would probably care about order.

Maybe it's better to override `==` for `AbstractOptimalEdgeSet` instead but that leads to the issue that we wouldn't be able to construct instances for unit tests (maybe, let's make constructors for the concrete subtypes?)

Reply:
Yes, this overrides `==` for any 

```julia
a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}}
```

However, I think we will be ok... We could always define custom types to represent what we want. For example, 

```julia
abstract type AbstractGraphPath{T <: Int} end

struct GraphPath{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
end

struct GraphTrail{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
    # Additional properties or methods specific to trails
end

struct GraphWalk{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
    # Additional properties or methods specific to walks
end
```

For each of these structures, we could add methods to construct them from a given graph, ensuring that they follow their respective constraints (e.g., no repeated edges in a trail). Here's an example **constructor** for `GraphPath`:

```julia
function GraphPath(vertices::Vector{Int}, edges::Vector{Tuple{Int, Int}})
    # Check if the number of vertices and edges are consistent.
    if length(vertices) - 1 != length(edges)
        error("Number of vertices and edges are not consistent")
    end

    # Validate that edges correctly connect the vertices.
    for (i, edge) in enumerate(edges)
        if edge != (vertices[i], vertices[i+1]) && edge != (vertices[i+1], vertices[i])
            error("Edge $edge does not connect vertices $(vertices[i]) and $(vertices[i+1])")
        end
    end

    # Additional checks like no repeated vertices or edges for a path
    if length(unique(vertices)) != length(vertices) || length(unique(edges)) != length(edges)
        error("Invalid path: repeated vertices or edges")
    end

    return GraphPath(vertices, edges)
end
```

Then we could write a new method for `==` 

```julia 
==(
        a::Union{Vector{GraphPath, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphPath, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 

==(
        a::Union{Vector{GraphTrail, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphTrail, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 

==(
        a::Union{Vector{GraphWalk, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphWalk, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 
```

However, we would want to be really careful if we did something like this and it **would most definitely be a separate PR if we decide to do this**.

Reply:
I'll leave it up to you to decide what is best 🙂